docs/design/datacontracts/Metadata.md

```diff
@@ -0,0 +1,287 @@
+# Contract Metadata
+
+This contract is for exploring the properties of the metadata of values on the heap or on the stack in a .NET process.
+
+## APIs of contract
+
+A `MethodTable` is the runtime representation of the type information about a value.  Given a `TargetPointer` address, the `Metadata` contract provides a `MethodTableHandle` for querying the `MethodTable`.
+
+``` csharp
+struct MethodTableHandle
+{
+    // no public properties or constructors
+
+    internal TargetPointer Address { get; }
+}
+```
+
+``` csharp
+    #region MethodTable inspection APIs
+    public virtual MethodTableHandle GetMethodTableHandle(TargetPointer targetPointer) => throw new NotImplementedException();
+
+    public virtual TargetPointer GetModule(MethodTableHandle methodTable) => throw new NotImplementedException();
+    // A canonical method table is either the MethodTable itself, or in the case of a generic instantiation, it is the
+    // MethodTable of the prototypical instance.
+    public virtual TargetPointer GetCanonicalMethodTable(MethodTableHandle methodTable) => throw new NotImplementedException();
+    public virtual TargetPointer GetParentMethodTable(MethodTableHandle methodTable) => throw new NotImplementedException();
+
+    public virtual uint GetBaseSize(MethodTableHandle methodTable) => throw new NotImplementedException();
+    // The component size is only available for strings and arrays.  It is the size of the element type of the array, or the size of an ECMA 335 character (2 bytes)
+    public virtual uint GetComponentSize(MethodTableHandle methodTable) => throw new NotImplementedException();
+
+    // True if the MethodTable is the sentinel value associated with unallocated space in the managed heap
+    public virtual bool IsFreeObjectMethodTable(MethodTableHandle methodTable) => throw new NotImplementedException();
+    public virtual bool IsString(MethodTableHandle methodTable) => throw new NotImplementedException();
+    // True if the MethodTable represents a type that contains managed references
+    public virtual bool ContainsPointers(MethodTableHandle methodTable) => throw new NotImplementedException();
+    public virtual bool IsDynamicStatics(MethodTableHandle methodTable) => throw new NotImplementedException();
+    public virtual ushort GetNumMethods(MethodTableHandle methodTable) => throw new NotImplementedException();
+    public virtual ushort GetNumInterfaces(MethodTableHandle methodTable) => throw new NotImplementedException();
+    public virtual ushort GetNumVirtuals(MethodTableHandle methodTable) => throw new NotImplementedException();
+    public virtual ushort GetNumVtableSlots(MethodTableHandle methodTable) => throw new NotImplementedException();
```

Comment:
I would question the SOS need to print these numbers. These numbers go into the MethodTable data layout details. Are the number of slots actually needed by anything else except some SOS command printing them?

Reply:
We discussed this a little bit in our standup.  The explanation was that some of these fields, while they're only used for SOS command output are important for a number of reasons:

1. Some 3rd party tooling scrapes SOS output. So eliminating the slot counts from the output entirely or changing the ouptut format too drastically may end up breaking those uses
2. The actual values are useful for engineers working on bringing up the runtime on new platforms

Reply:
I am sorry I was not able to make it to the standup today.

> Some 3rd party tooling scrapes SOS output.

I do not see a problem with that. The screen scrapping done by 3rd party tools should not prevent us from making the SOS better.

> The actual values are useful for engineers working on bringing up the runtime on new platforms

The main value of SOS dump commands - for both platform bring up and general debugging - is in displaying information that is hard (labor intensive) to get manually via regular windbg/lldb commands. Type name is an example of such information.

NumVtableSlots is easy to get using regular windbg/lldb commands, only needed to interpret data layout of the methodtable, and rarely used for low-level debugging. It does not need to be part of Dump... output. I would even say that Dump... would be better without it.. We should be free to delete it to make SOS better. https://github.com/dotnet/coreclr/pull/16560 or https://github.com/dotnet/diagnostics/pull/4751 are examples of changes like that.

Reply:
> I do not see a problem with that. The screen scrapping done by 3rd party tools should not prevent us from making the SOS better.

I agree with this in the long term and how we should frame the ultimate outcome of the cDAC workstream. For these initial phases though I tend to consider them more proving this can be done in a seamless and transparent manner. This implies ensuring we maintain the illusion of the static nature should be a priority until we can announce changes in this space and start to innovate. Basically, for now I think we should match precise output unless there is a compelling or hard-blocking reason to deviate from that.

Reply:
> for now I think we should match precise output unless there is a compelling or hard-blocking reason to deviate from that.

Are you saying that we should revert https://github.com/dotnet/runtime/commit/32665b21780fbb54834457d9f1b918b1dc5ea9cf to match existing output    and that we should avoid [cleaning up clutter in the output of SOS commands](https://github.com/dotnet/diagnostics/pull/4751#issuecomment-2184315458)?

It does not sound right to me. I think we should take every opportunity to make small incremental simplifications as we run into them.

Reply:
> Are you saying that we should revert https://github.com/dotnet/runtime/commit/32665b21780fbb54834457d9f1b918b1dc5ea9cf to match existing output and that we should avoid https://github.com/dotnet/diagnostics/pull/4751#issuecomment-2184315458? Both of these changes can break screen scraping tools.

As an initial goal, I would've said "yes, we should stick to avoiding breaking changes in this space". However, it seems like that ship has already sailed. I want to innovate in this space a great deal so as an end goal I agree. My concern was and is the burden this can place on automation that relies on scapping. We end up encouraging odd solutions as people try to anticipate or adapt to our "infight work". Since it seems that isn't a current goal, I agree then that we should remove as much cruft as possible.

Reply:
Perhaps there should be an SOS query that can help scrapping tooling know what they can expect? Is a new SOS command along the lines of `!sos.version` appropriate in this case? I truly don't know how onerous this is so I could be concerned about something that has limited impact. Perhaps @leculver can shed some light on a version command option, if one doesn't already exist.

Reply:
Updated the current PR to always set `DacpMethodTableData::wNumVirtuals` and `DacpMethodTableData::wNumVtableSlots` to 0 in both the cDAC and the Legacy DAC.   Updated the Metadata contract to exclude getters for this info.

Reply:
Related change in dotnet/diagnostics to not print the counts if they're 0 https://github.com/dotnet/diagnostics/pull/4760

I checked and https://github.com/Microsoft/clrmd doesn't use these fields. They only look at `wNumMethods`

Reply:
> Perhaps there should be an SOS query that can help scrapping tooling know what they can expect? Is a new SOS command along the lines of `!sos.version` appropriate in this case? I truly don't know how onerous this is so I could be concerned about something that has limited impact. Perhaps @leculver can shed some light on a version command option, if one doesn't already exist.

Sorry I missed this ping.

My guiding principle is:  When possible, and when it doesn't cause too much work, don't regress or take back features that affect debugging.  That includes .Net Core and Desktop Framework (since we can debug 4.5+ desktop framework with current SOS).

That's my only real request.

> Perhaps there should be an SOS query that can help scrapping tooling know what they can expect? Is a new SOS command along the lines of !sos.version appropriate in this case? 

I don't think we need that.  **The output of SOS does not constitute an interface.**  We shouldn't bend over backwards to accommodate screen scraping.  Literally the original reason ClrMD was literally built was to replace screen-scraping SOS.  We don't have all functionality built into ClrMD, but the SOS source is open-source and folks can reimplement functionality from there if they need to.

I haven't fully dug into everything here, but I'm fine with these changes.  I'd honestly be fine if folks wanted to randomize the output of SOS to _intentionally_ break screen scraping.  (Hyperbole.  Obviously, we would never actually do that.)  I owned SOS 15 years ago and built a lot of its functionality.  Screen scraping was the bane of my existence because we couldn't do _anything_ to SOS in the presence of screen scraping.  Just use ClrMD if you need that programmatic access!  No screen scraping!